fortran/Makefile

```diff
@@ -1,66 +1,75 @@
-FCPREFLAGS  := -cpp
-FCCUDAFLAGS :=
-FCFLAGS     := -ffree-line-length-none
-CUDAFORDEP  := libcudafor
-CUDALINK    := -L$(CUDA_LIB) -lcudart
-CUDAFORLINK := -lcudafor
-else ifeq ($(FC), ifort)
-FCMODFLAGS  := -module $(INCDIR)
-FCPREFLAGS  := -fpp
-FCCUDAFLAGS :=
-FCFLAGS     :=
-CUDAFORDEP  := libcudafor
-CUDALINK    := -L$(CUDA_LIB) -lcudart
-CUDAFORLINK := -lcudafor
-else
+LIBNAME    := libncclfor.so
+LIBSONAME  := $(patsubst %,%.$(NCCL_MAJOR),$(LIBNAME))
+LIBTARGET  := $(patsubst %,%.$(NCCL_MAJOR).$(NCCL_MINOR).$(NCCL_PATCH),$(LIBNAME))
+LIBLINK    += $(patsubst lib%.so,-l%,$(LIBNAME))
+
+LIBCUDAFOR := libcudafor.so
+
+ifeq ($(FCNAME), pgf90)
 FCMODFLAGS  := -module $(INCDIR)
 FCPREFLAGS  := -Mpreprocess
 FCCUDAFLAGS := -Mcuda,cuda$(CUDA_MAJOR).$(CUDA_MINOR)
 FCFLAGS     := -fast -O3
-CUDAFORDEP  :=
-CUDALINK    :=
-CUDAFORLINK :=
+else
+# non-pgf90 do not have CUDA support, compile our own CUDA lib
+CUDAFORDEP  := $(LIBDIR)/$(LIBCUDAFOR) 
+CUDALINK    := -L$(CUDA_LIB) -lcudart
+CUDAFORLINK := -lcudafor
+ifeq ($(FCNAME), gfortran)
+FCMODFLAGS  := -J$(INCDIR)
+FCPREFLAGS  += -cpp
+FCFLAGS     += -ffree-line-length-none
+else ifeq ($(FC), ifort)
+FCMODFLAGS  := -module $(INCDIR)
+FCPREFLAGS  += -fpp
+endif
+endif
+
+ifeq ($(VERBOSE), 0)
+.SILENT:
 endif
 
 lib: $(CUDAFORDEP)
-	$(FC) -c $(FCMODFLAGS) $(FCPREFLAGS) -fPIC $(FCCUDAFLAGS) $(FCFLAGS) src/ncclfor.f90 -o $(OBJDIR)/ncclfor.o
-	$(FC) -shared -Wl,--no-as-needed -Wl,-soname,$(LIBSONAME) $(OBJDIR)/ncclfor.o -o $(LIBDIR)/$(LIBTARGET)
+	make $(LIBDIR)/$(LIBTARGET)
+
+$(LIBDIR)/$(LIBTARGET): $(OBJDIR)/ncclfor.o
+	@printf "Linking   %-25s\n" $@
+	$(FC) -shared -Wl,--no-as-needed -Wl,-soname,$(LIBSONAME) $< -o $(LIBDIR)/$(LIBTARGET)
 	ln -sf $(LIBSONAME) $(LIBDIR)/$(LIBNAME)
 	ln -sf $(LIBTARGET) $(LIBDIR)/$(LIBSONAME)
 
-export
+$(LIBDIR)/$(LIBCUDAFOR): $(OBJDIR)/cudafor.o
+	@printf "Linking   %-25s\n" $@
+	$(FC) -shared -Wl,--no-as-needed -Wl,-soname,libcudafor.so $< -o $(LIBDIR)/$(LIBCUDAFOR)
 
-libcudafor:
-	$(MAKE) -C cudafor lib
+$(OBJDIR)/%.o: src/%.f90 
+	@printf "Building  %-25s > %-24s\n" $< $@
+	$(FC) -c $(FCMODFLAGS) $(FCPREFLAGS) -fPIC $(FCCUDAFLAGS) $(FCFLAGS) $< -o $@
 
 TESTS := reduce_ptr_out allreduce_ptr_out reducescatter_ptr_out broadcast_ptr allgather_ptr_out
-ifneq ($(FC), gfortran)
-ifneq ($(FC), ifort)
+ifeq ($(FCNAME), pgf90)
 TESTS += reduce_arr_out allreduce_arr_out reducescatter_arr_out broadcast_arr allgather_arr_out
 endif
-endif
 
 TESTDIR  := $(BUILDDIR)/test/fortran
 TESTBINS := $(patsubst %,$(TESTDIR)/%,$(TESTS))
 
 test: lib $(TESTBINS)
 
-$(TESTDIR)/%: test/%.f90
+$(TESTDIR)/%: test/%.f90 lib
+	@printf "Building  %-25s > %-24s\n" $< $@
 	@mkdir -p $(TESTDIR)
 	$(FC) $(FCCUDAFLAGS) $(FCFLAGS) $< $(CUDALINK) -I$(INCDIR) -L$(LIBDIR) $(CUDAFORLINK) $(LIBLINK) -o $@
 
 clean:
-	rm -rf $(INCDIR)/*.mod
-	rm -rf $(LIBDIR)/lib*for.so*
-	rm -rf $(OBJDIR)/*for.o
+	rm -f $(LIBDIR)/$(LIBTARGET) $(LIBDIR)/$(LIBCUDAFOR) $(OBJDIR)/*for.o $(INCDIR)/*.mod
```

Comment:
Changing `$(LIBDIR)/lib*for.so*` to `$(LIBDIR)/$(LIBTARGET)` doesn't delete the `$(LIBDIR)/libncclfor.so` and `$(LIBDIR)/libncclfor.so.1` links.